### PR TITLE
Record an informative stack trace & telemetry whenever `FilePath` agent-to-controller access is permitted

### PR DESCRIPTION
**Pull request by jglick**
Follows up #5880. Complements #5884. Obsoleted by #5885.

Example in system log:

```
WARNING	j.s.s2m.DefaultFilePathFilter$1#op: Permitting agent-to-controller read on …/test/target/j h2390377948147974035/userContent/target_public.txt
Command UserRequest:hudson.FilePath$ReadToString@191a1735 created at
	at hudson.remoting.Command.<init>(Command.java:81)
	at hudson.remoting.Request.<init>(Request.java:113)
	at hudson.remoting.Request.<init>(Request.java:108)
	at hudson.remoting.UserRequest.<init>(UserRequest.java:77)
	at hudson.remoting.Channel.call(Channel.java:999)
	at hudson.FilePath.act(FilePath.java:1159)
	at hudson.FilePath.act(FilePath.java:1148)
	at hudson.FilePath.readToString(FilePath.java:2280)
	at jenkins.security.s2m.AdminFilePathFilterTest$ReadFileS2MCallable.call(AdminFilePathFilterTest.java:111)
	at jenkins.security.s2m.AdminFilePathFilterTest$ReadFileS2MCallable.call(AdminFilePathFilterTest.java:103)
	at hudson.remoting.UserRequest.perform(UserRequest.java:211)
	at hudson.remoting.UserRequest.perform(UserRequest.java:54)
	at hudson.remoting.Request$2.run(Request.java:376)
	at hudson.remoting.InterceptingExecutorService.lambda$wrap$0(InterceptingExecutorService.java:78)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```

Note that the stack of the `MasterToSlaveCallable` construction (i.e., plugin code running on the controller side initiating everything) is not recorded here—that gets attached only later, if an exception is actually thrown—but this is probably good enough since it should identity lines of plugin code responsible. (Anyway you would have no such stack trace if there were background threads from the agent accessing controller `FilePath`s.)

Telemetry example after doing a couple builds using the `log-parser` plugin and running a system script to call `createContent()`:

```console
$ echo '{"traces":["Command UserRequest:hudson.FilePath$CopyTo@66b2627a created at\n\tat hudson.remoting.Command.<init>(Command.java:81)…(Thread.java:748)\n","Command UserRequest:hudson.FilePath$CopyTo@7c8ee7be created at\n…"]}' | jq -r '.traces[0]'
Command UserRequest:hudson.FilePath$CopyTo@66b2627a created at
	at hudson.remoting.Command.<init>(Command.java:81)
	at hudson.remoting.Request.<init>(Request.java:113)
	at hudson.remoting.Request.<init>(Request.java:108)
	at hudson.remoting.UserRequest.<init>(UserRequest.java:77)
	at hudson.remoting.Channel.call(Channel.java:999)
	at hudson.FilePath.act(FilePath.java:1165)
	at hudson.FilePath.act(FilePath.java:1154)
	at hudson.FilePath.copyTo(FilePath.java:2485)
	at hudson.FilePath.copyTo(FilePath.java:2440)
	at hudson.plugins.logparser.LogParserStatusComputer.computeStatusMatches(LogParserStatusComputer.java:88)
	at hudson.plugins.logparser.LogParserStatusComputer.access$000(LogParserStatusComputer.java:22)
	at hudson.plugins.logparser.LogParserStatusComputer$1.call(LogParserStatusComputer.java:54)
	at hudson.plugins.logparser.LogParserStatusComputer$1.call(LogParserStatusComputer.java:47)
	at hudson.remoting.UserRequest.perform(UserRequest.java:211)
	at hudson.remoting.UserRequest.perform(UserRequest.java:54)
	at hudson.remoting.Request$2.run(Request.java:376)
	at hudson.remoting.InterceptingExecutorService.lambda$wrap$0(InterceptingExecutorService.java:78)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```

(**Edit**: after e9eb403709 there should be just one entry.)

Telemetry help:
telemetry-help

### Proposed changelog entries

* Issue a warning to the system log when using agent-to-controller file manipulation idioms considered for deprecation, and collect telemetry about this as well.

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

### CODE REVIEW

File: core/src/main/java/jenkins/security/s2m/DefaultFilePathFilter.java
Comment (jglick) [inline core/src/main/java/jenkins/security/s2m/DefaultFilePathFilter.java:69]: Good enough for now; could be cleaned up by introducing a proper API in `remoting`. I am not inclined to bother if #5885 is under consideration anyway.

**Comment by daniel-beck** [comment]
> Could be extended to implement [JEP-214 telemetry] if desired.

Seems like a more appropriate approach here as admins are very rarely the ones able to address such issues. "Please report a bug", especially with fairly severe log spam on a busy instance, seems inferior to just collecting these on our side and reviewing every few weeks. I expect this should be fairly low effort through restoring some of the Java 11 telemetry code and adapting it.

File: core/src/main/java/jenkins/security/s2m/DefaultFilePathFilter.java
Comment (daniel-beck) [inline core/src/main/java/jenkins/security/s2m/DefaultFilePathFilter.java:70]:
```suggestion
                        LOGGER.log(Level.WARNING, "Permitting agent-to-controller '" + op + "' on '" + f  + "' which will be removed soon. See https://www.jenkins.io/redirect/permitted-agent-to-controller-file-access", (Throwable) createdAt.get(((ThreadLocal) current.get(null)).get()));
```

Makes this actionable to admins. We'll need to provide some instructions how to identify the culprit, best done on the site, and ask that a bug be reported.

File: core/src/main/java/jenkins/security/s2m/DefaultFilePathFilter.java
Comment (daniel-beck) [inline core/src/main/java/jenkins/security/s2m/DefaultFilePathFilter.java:72]: We know both `op` and `f` here, so a better error message would probably be useful? Unless we expect this to never happen?

```suggestion
                        LOGGER.log(Level.WARNING, "Permitting agent-to-controller '" + op + "' on '" + f + "', but failed to identify the caller. See https://www.jenkins.io/redirect/permitted-agent-to-controller-file-access", x);
```

**Review by daniel-beck — APPROVED** [review]
(Both suggestions untested.)

As a side benefit, if I screwed up the security fixes, at least these messages help identify what happened 😅

File: core/src/main/java/jenkins/security/s2m/DefaultFilePathFilter.java
Comment (jglick) [reply core/src/main/java/jenkins/security/s2m/DefaultFilePathFilter.java:70]: https://www.jenkins.io/redirect/permitted-agent-to-controller-file-access is a 404.

File: core/src/main/java/jenkins/security/s2m/DefaultFilePathFilter.java
Comment (jglick) [reply core/src/main/java/jenkins/security/s2m/DefaultFilePathFilter.java:72]: If you get an error _here_, the `file` and `op` are beside the point—something bigger is broken.

**Comment by jglick** [comment]
> Seems like a more appropriate approach here

Do you think we should hold this up to add telemetry, or merge as is and then consider a follow-up telemetry PR?

**Comment by jglick** [comment]
BTW @daniel-beck https://github.com/jenkinsci/jenkins/blob/090779fb4e88f8105a1cde0e2eaca725e7b3966d/core/src/main/java/jenkins/telemetry/impl/SecuritySystemProperties.java#L77-L97 has not been updated in two years…

File: core/src/main/java/jenkins/telemetry/impl/SlaveToMasterFileCallableUsage.java
Comment (jglick) [inline core/src/main/java/jenkins/telemetry/impl/SlaveToMasterFileCallableUsage.java:61]: Just making this up. Frankly I have never understand why `Telemetry` defines start & end dates. Would it not be simpler for it to just be unconditionally active, and when we no longer want it, we delete it from sources?

Comment (jglick) [reply core/src/main/java/jenkins/telemetry/impl/SlaveToMasterFileCallableUsage.java:61]: Though, maybe we want this telemetry backported to LTS so as to get broader usage information, especially from admins who are more reluctant to upgrade stuff?

File: core/src/main/resources/jenkins/telemetry/impl/SlaveToMasterFileCallableUsage/description.jelly
Comment (daniel-beck) [inline core/src/main/resources/jenkins/telemetry/impl/SlaveToMasterFileCallableUsage/description.jelly:6]:
```suggestion
    and <a href="https://www.jenkins.io/security/advisory/2021-11-04/" target="_blank" rel="noopener noreferrer">recently</a>
```

File: core/src/main/java/jenkins/telemetry/impl/SlaveToMasterFileCallableUsage.java
Comment (daniel-beck) [reply core/src/main/java/jenkins/telemetry/impl/SlaveToMasterFileCallableUsage.java:61]:
```suggestion
        return LocalDate.of(2022, 3, 1);
```

~3 months once merged seems reasonable, assuming we backport. This should catch lesser-used plugins with some consistency.

File: core/src/main/resources/jenkins/telemetry/impl/SlaveToMasterFileCallableUsage/description.jelly
Comment (daniel-beck) [inline core/src/main/resources/jenkins/telemetry/impl/SlaveToMasterFileCallableUsage/description.jelly:12]: Vague statement is vague.

File: core/src/main/java/jenkins/security/s2m/DefaultFilePathFilter.java
Comment (daniel-beck) [inline core/src/main/java/jenkins/security/s2m/DefaultFilePathFilter.java:74]:
```suggestion
                        LOGGER.log(Level.WARNING, "Permitting agent-to-controller '" + op + "' on '" + f + "'. This is deprecated and will soon be rejected. Learn more: https://www.jenkins.io/redirect/permitted-agent-to-controller-file-access", trace);
```

Give admins a chance to understand why this is a warning and why they should care.

**Review by daniel-beck — APPROVED** [review]
Looks good, thanks.

----

I manually tested this and confirmed that content is submitted to uplink.

Agent script console:
```
new FilePath(jenkins.agents.AgentComputerUtil.getChannelToMaster(), "/root/.jenkins/userContent/readme.txt").readToString()
```

Then System script console:
```
ExtensionList.lookupSingleton(jenkins.telemetry.Telemetry.TelemetryReporter.class).run()
```

----

Something we might consider: Discard everything that has entries like
```
Script1.run(Script1.groovy:1)\n\tat groovy.lang.GroovyShell.evaluate(GroovyShell.java:585)
```
in its stack trace? They're not useful and likely spammy (many different stack traces).

Additionally, if we care about lines, we may consider collecting Jenkins version and plugin versions? Unsure whether this would be helpful wrt identifying incomplete fixes and such, but probably not needed?

File: core/src/main/java/jenkins/security/s2m/DefaultFilePathFilter.java
Comment (jglick) [reply core/src/main/java/jenkins/security/s2m/DefaultFilePathFilter.java:74]: This is a 404.

File: core/src/main/resources/jenkins/telemetry/impl/SlaveToMasterFileCallableUsage/description.jelly
Comment (jglick) [reply core/src/main/resources/jenkins/telemetry/impl/SlaveToMasterFileCallableUsage/description.jelly:12]: Vague in what way? What would you suggest?

**Comment by jglick** [comment]
> Discard everything that has entries like
> 
> ```
> Script1.run(Script1.groovy:1)\n\tat groovy.lang.GroovyShell.evaluate(GroovyShell.java:585)
> ```
> 
> in its stack trace?

I could, but would anyone besides you be trying to run this sort of thing from the script console?

**Comment by jglick** [comment]
> we may consider collecting Jenkins version and plugin versions

Surprised to learn this is not already collected by telemetry as a generic bit of context. Implementations should not need to do things like https://github.com/jenkinsci/jenkins/blob/090779fb4e88f8105a1cde0e2eaca725e7b3966d/core/src/main/java/jenkins/telemetry/impl/StaplerDispatches.java#L87-L98 because that should be part of https://github.com/jenkinsci/jenkins/blob/090779fb4e88f8105a1cde0e2eaca725e7b3966d/core/src/main/java/jenkins/telemetry/Telemetry.java#L195-L198 for every submission.

> helpful wrt identifying incomplete fixes and such

Seems doubtful. When fixing a S2MFC usage I would typically just be deleting the affected code and rewriting it. So it should be obvious enough that the stack trace is coming from obsolete code.

File: core/src/main/java/jenkins/security/s2m/DefaultFilePathFilter.java
Comment (daniel-beck) [reply core/src/main/java/jenkins/security/s2m/DefaultFilePathFilter.java:74]: Will be a new page once this is merged.

Comment (daniel-beck) [reply core/src/main/java/jenkins/security/s2m/DefaultFilePathFilter.java:74]: It doesn't, but we can easily create it. Doesn't make a lot of sense to create it before this change gets accepted (of course, we can prepare the corresponding PR). Could be as simple as redirecting to this PR, or my JEP PR, or the JEP once merged, to provide context.

#5885 links to `https://www.jenkins.io/redirect/AdminWhitelistRule` which doesn't exist yet either 🤷

**Review by daniel-beck — APPROVED** [review]

**Comment by daniel-beck** [comment]
FWIW this only captures `FilePath` methods, because those are the only `SlaveToMasterFileCallable` implementations that use `FilePathFilter`.

This matches what https://plugins.jenkins.io/remoting-security-workaround/ blocks and https://github.com/jenkinsci/jep/pull/381 proposes for removal.

`SlaveToMasterFileCallable` in general, while proposed for deprecation by https://github.com/jenkinsci/jenkins/pull/5884, is not caught here (but those are easy to identify statically).

**Comment by daniel-beck** [comment]
`spotless` complains about imports.

File: test/src/test/java/jenkins/security/s2m/AdminFilePathFilterTest.java
Comment (daniel-beck) [inline test/src/test/java/jenkins/security/s2m/AdminFilePathFilterTest.java:51]: Spotless expects static imports on top.

Comment (jglick) [reply test/src/test/java/jenkins/security/s2m/AdminFilePathFilterTest.java:51]: :angry:

Comment (daniel-beck) [reply test/src/test/java/jenkins/security/s2m/AdminFilePathFilterTest.java]: I know, I took care of that for you in the private PRs because we had a deadline, but here you're on your own 😛

**Comment by jglick** [comment]
Anyone else able to review? Would love to have this in the next weekly.

**Review by basil — APPROVED** [review]
This PR is now ready for merge. We will merge it after approximately 24 hours if there is no negative feedback. Please see the [merge process documentation](https://github.com/jenkinsci/jenkins/blob/master/docs/MAINTAINERS.adoc#merge-process) for more information about the merge process. Thanks!

**Comment by basil** [comment]
BTW I have wanted to gather telemetry like this myself in a few cases but been hesitant because I wasn't sure how to gather it and look at the results. Any documentation for this subsystem would be highly appreciated!

**Comment by jglick** [comment]
@daniel-beck https://www.jenkins.io/redirect/permitted-agent-to-controller-file-access is still a 404. When will this be implemented?

**Comment by daniel-beck** [comment]
> BTW I have wanted to gather telemetry like this myself in a few cases but been hesitant because I wasn't sure how to gather it and look at the results. Any documentation for this subsystem would be highly appreciated!

This system is documented at https://github.com/jenkinsci/jep/blob/master/jep/214/README.adoc

To access submissions, we grant access on request to `uplink.jenkins.io`. (Your telemetry should have a reasonable use case, but anything that makes it into core should be fine.)

**Pull request merged**
